File: leetcode/python/canThreePartsEqualSum.py
from typing import List


class Solution:
    def canThreePartsEqualSum(self, arr: List[int]) -> bool:
        sum_arr = sum(arr)
        if sum_arr % 3 != 0:
            return False

        target, temp, res = sum_arr // 3, 0, 0

        for num in arr:
            temp += num

            if temp == target:
                res += 1
                temp = 0

        return res >= 3


print(Solution().canThreePartsEqualSum([0, 2, 1, -6, 6, -7, 9, 1, 2, 0, 1]))
print(Solution().canThreePartsEqualSum([0, 2, 1, -6, 6, 7, 9, -1, 2, 0, 1]))
print(Solution().canThreePartsEqualSum([3, 3, 6, 5, -2, 2, 5, 1, -9, 4]))

File: leetcode/js/canThreePartsEqualSum.js
/**
 * @param {number[]} arr
 * @return {boolean}
 */
const canThreePartsEqualSum = function (arr) {
  const target = arr.reduce((a, b) => a + b) / 3;
  let sum = 0;
  let temp = 0;

  for (const num of arr) {
    temp += num;
    if (temp === target) {
      sum++;
      temp = 0;
    }
  }
  return sum >= 3;
};

console.log(canThreePartsEqualSum([0, 2, 1, -6, 6, -7, 9, 1, 2, 0, 1]));
console.log(canThreePartsEqualSum([0, 2, 1, -6, 6, 7, 9, -1, 2, 0, 1]));
console.log(canThreePartsEqualSum([3, 3, 6, 5, -2, 2, 5, 1, -9, 4]));
console.log(canThreePartsEqualSum([0, 0, 0, 0]));
console.log(canThreePartsEqualSum([1, -1, 1, -1]));
